Core/Src/main.c

```diff
@@ -1,181 +1,249 @@
-/**
-  * @brief  Reports the name of the source file and the source line number
-  *         where the assert_param error has occurred.
-  * @param  file: pointer to the source file name
-  * @param  line: assert_param error line source number
-  * @retval None
-  */
-void assert_failed(uint8_t *file, uint32_t line)
-{
-  /* USER CODE BEGIN 6 */
-  /* User can add his own implementation to report the file name and line number,
-     ex: printf("Wrong parameters value: file %s on line %d\r\n", file, line) */
-  /* USER CODE END 6 */
-}
-#endif /* USE_FULL_ASSERT */
-
-/************************ (C) COPYRIGHT STMicroelectronics *****END OF FILE****/
+/* USER CODE BEGIN Header */
+/**
+  ******************************************************************************
+  * @file           : main.c
+  * @brief          : Main program body
+  ******************************************************************************
+  * @attention
+  *
+  * <h2><center>&copy; Copyright (c) 2021 STMicroelectronics.
+  * All rights reserved.</center></h2>
+  *
+  * This software component is licensed by ST under BSD 3-Clause license,
+  * the "License"; You may not use this file except in compliance with the
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+#define ADC_VAL_TO_COUNTS 46.875  
+#define ADC_MIN_COUNTS_DUTY_CYCLE 48000
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+// Is extern from spi.c/h
+extern SPI_HandleTypeDef hspi1;
+// Is extern from tim.c/h
+extern TIM_HandleTypeDef htim2;
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+  uint8_t spi_transmit_data_buf, spi_receive_data_buf = 0;

```

Comment:
So for these two buffers, since we already know that we are going to be transmitting and receiving 3 bytes of data based on this figure in the data sheet (part i highlighted in red). Why not just make an array of 3 for both (where the transmit buffer is const since it wouldn't change) and then using the transmit and receive function for all 3 bytes? 

![image](https://user-images.githubusercontent.com/69175242/213580318-9e76c780-ccb2-42b2-ac1d-2b44f865c39d.png)

Reply:
I chose to break it up into chunks to have better error handling of the 3-byte exchange. Instead of doing the entire exchange in one 'go' (may or may not be how it happens in hardware), I wanted to send the first two byte seperate from the last one. This way, if there's something wrong we can catch it earlier in the exchange (By verifying the NULL bit) and go back to the start. What are your thoughts on this approach? Am I sacrificing potentional performance by going piecemeal?

Reply:
I see. In this case I think it is totally fine bec the application isn't that large, but of course it is application specific, but in general actually yes, if you can add some type of error checking, that is very good! I do like the different approach!